### PR TITLE
Add NIST KAT check for KEMs

### PR DESCRIPTION
**Pull request by dstebila**
This addresses #110. It adds a copy of the NIST KAT generation program, then checks that the outputs of that KAT program match a hash value.  The hash value should be generated directly from the KAT files in the original NIST submission.  

Note that it is currently only implemented for KEMs.  If people are happy with this, then we can add it for signatures.  

Note as well that the KAT values for our kyber768 implementation don't match the ones in the Round 2 submission.  If memory serves, this is because kyber768 was modified for Round after the version that was added here.  I added a line to temporarily skip this test for kyber768.

#### Manually checked properties

* [ ] No stringification macros
* [ ] Output-parameter pointers in functions are on the left
* [ ] Negative return values on failure of API functions (within restrictions of FO transform).
* [ ] `const` arguments are labeled as `const`
* [ ] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [ ] All integer types are of fixed size, using `stdint.h` types (including `uint8_t` instead of `unsigned char`)
  * [ ] Integers used for indexing are of size `size_t`

### CODE REVIEW

File: test/crypto_kem/nistkat.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    nist_kat_init(seed, NULL, 256);
+
+    rc = crypto_kem_keypair(public_key, secret_key);
+    if (rc != 0) {
+        fprintf(stderr, "[kat_kem] %s ERROR: crypto_kem_keypair failed!\n", CRYPTO_ALGNAME);
+        return -1;
+    }
+    fprintBstr(fh, "pk = ", public_key, CRYPTO_PUBLICKEYBYTES);
+    fprintBstr(fh, "sk = ", secret_key, CRYPTO_SECRETKEYBYTES);
```
Comment (thomwiggers) [inline test/crypto_kem/nistkat.c]: I see that the NIST KAT files still contain the secret keys, thus not solving the problems discussed in #93. Does the (original) KAT format address that problem at all? Or do we still need to think about a new format that specifies equivalence up to secret keys (whenever that issue comes up)

File: common/nistkatrng.c [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+//    key - 256-bit AES key
+//    ctr - a 128-bit plaintext value
+//    buffer - a 128-bit ciphertext value
+static void AES256_ECB(uint8_t *key, uint8_t *ctr, uint8_t *buffer) {
+    aes256ctx ctx;
+    aes256_keyexp(&ctx, key);
+    aes256_ecb(buffer, ctr, 1, &ctx);
+}
+
+void nist_kat_init(uint8_t *entropy_input, const uint8_t *personalization_string, int security_strength);
```
Comment (thomwiggers) [inline common/nistkatrng.c]: is this prototype necessary?

File: test/crypto_kem/nistkat.c (hunk 1 above)
Comment (joostrijneveld) [reply test/crypto_kem/nistkat.c]: The original KAT format 'addresses' that 'problem' in the sense that it includes secret keys. By requiring consistency with those files, I suppose #93 is forced towards doing the same thing.

Comment (thomwiggers) [reply test/crypto_kem/nistkat.c]: Yeah, Matthias and I talked a bit about it at the coffee machine. The problem isn't limited to secret keys, but also for example the choice of randomness in encapsulation. As identified in #93, I guess that the question indeed is where we draw the line in specifications: do they include these kinds of details?

So I guess the NIST format also doesn't answer any of these questions and any equivalence testing could really only be interoperability of implementations (i.e. do they decapsulate to the same shared secret key).

File: common/nistkatrng.c (hunk 2 above)
Comment (joostrijneveld) [reply common/nistkatrng.c]: Yes, because of `-Wmissing-prototypes` (see #108 and #109)

**Comment by thomwiggers** [comment]
I'm still a bit unclear on how the NIST format works, didn't it have a request/response type of structure? I'm only seeing one side of it here.

**Comment by dstebila** [comment]
> I'm still a bit unclear on how the NIST format works, didn't it have a request/response type of structure? I'm only seeing one side of it here.

I think the request/response structure can be used if you want to write an interactive program that receives the seed inputs, then generates responses, and compare them against the responses.  The way I've implemented here is that it starts from the same master seed, then checks that it generates the same byte-for-byte output as in the response file.

**Pull request merged**